# Hacker News Project !

I will be comparing [hacker news](https://news.ycombinator.com/) posts. `User questions` are one half of the posts, and `user contributions` are the other half. This project will make an inference about which one recieves more comment responses on average.

Let us begin with reading the Hacker News data set.

In [1]:
from csv import reader
file = open('hacker_news.csv')
file_reader = reader(file)
hn = list(file_reader) #"hn" will represent the hacker news data set


In [2]:
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


Let's get rid of that header row, but first let me dislpay it. 

In [3]:
headers = hn[0]
print(headers)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


Now to get rid of that header row and display the first 5 rows of data.

In [4]:
hn = hn[1:] #This line of code is updating the the hacker news dataset by getting rid of the header row
print(hn[:5]) #This line is displaying the first 5 rows of the updated dataset

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


#### Now to separate the posts into 3 separate lists.

Going to start off by creating three empty lists. 

In [5]:
ask_posts = []
show_posts = []
other_posts = []

In [6]:
for row in hn:
    title = row[1]
    title = title.lower() #This line of code transformed all of the uppercase first letters in row 2 to lowercase
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

Now that we put all of the posts in separate lists, let's go ahead and count how many posts are in each list. 

In [7]:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


It makes sense that the `ask posts` have the  most comments, because people in the thread most likely ask and answer questions creating more comments.

Now to to display the first 5 rows of the ask and show posts. 

In [8]:
print(ask_posts[:5])
print('\n')
print(show_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 

Now lets see which one gets more posts on average. 

In [9]:
total_ask_comments = 0
comments_list = []
for x in ask_posts:
    num_comments = int(x[4])
    comments_list.append(num_comments)
total_ask_comments += sum(comments_list)

In [10]:
avg_ask_comments = total_ask_comments / len(comments_list)
print(avg_ask_comments)

14.038417431192661


In [11]:
total_show_comments = 0
comments_list = []
for x in show_posts:
    num_comments = int(x[4])
    comments_list.append(num_comments)
total_show_comments += sum(comments_list)

In [12]:
avg_show_comments = total_show_comments / len(comments_list)
print(avg_show_comments)

10.31669535283993


The data shows that [Hacker News](https://news.ycombinator.com/) recieves more "ask comments" than "show comments" on average.

Now let's see which hour attracts the most `ask posts`.

In [13]:
import datetime as dt
result_list = []
for element in ask_posts:
    created_at = element[6]
    num_comments = element[4]
    result_list.append([created_at, int(num_comments)])

    
print(result_list[:5])

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


Now to format and parse the time so that we're able to use the datetime class to determine which hour gets the most ask posts.

In [14]:
counts_by_hour = {}
comments_by_hour = {}
format_time = '%m/%d/%Y %H:%M'


for row in result_list:
    
    time_fr = row[0]
    comment = row[1]
   #This parsing and formatting goes crazy fr 
    time_fr = dt.datetime.strptime(time_fr, format_time).strftime('%H')
    
    if time_fr in counts_by_hour:
        counts_by_hour[time_fr] += 1
        comments_by_hour[time_fr] += comment
        
    else:
        counts_by_hour[time_fr] = 1
        comments_by_hour[time_fr] = comment

print(comments_by_hour)

    

{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


In [15]:
avg_by_hour = []
for time_fr in comments_by_hour:
    avg_by_hour.append([time_fr, comments_by_hour[time_fr] / counts_by_hour[time_fr]])
print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


Now to swap the order of the average and the hour.

In [16]:
swap_avg_by_hour = []
for time in avg_by_hour:
    swap_avg_by_hour.append([time[1], time[0]])
print(swap_avg_by_hour)

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [17]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print(sorted_swap[:5])

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21']]


Now let's display the top 5 hours for `ask posts`

In [18]:
print("Top 5 Hours for Ask Posts Comments")
for avg, hr in sorted_swap[:5]:
    print('{} : {:.2f}'.format(dt.datetime.strptime(hr, "%H").strftime("%H:%M"), avg))
    

Top 5 Hours for Ask Posts Comments
15:00 : 38.59
02:00 : 23.81
20:00 : 21.52
16:00 : 16.80
21:00 : 16.01


These are the the top 5 hours that ask posts are recieved by [Hacker News](https://news.ycombinator.com/). The data shows that  3:00p.m. gets the most comments on average at 38.59. 